core/src/main/java/org/elasticsearch/action/support/replication/ReplicatedOperation.java

```diff
@@ -0,0 +1,334 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public class ReplicatedOperation<Request extends ReplicationRequest<Request>, ReplicaRequest extends ReplicationRequest<ReplicaRequest>,
+        Response extends ReplicationResponse> {
+    final ESLogger logger;
+    final Request request;
+    final Supplier<ClusterState> clusterStateSupplier;
+    final String opType;
+    final AtomicInteger totalShards = new AtomicInteger();
+    final AtomicInteger pendingShards = new AtomicInteger();
+    final AtomicInteger successfulShards = new AtomicInteger();
+    final boolean executeOnReplicas;
+    final boolean checkWriteConsistency;
+    final Primary<Request, ReplicaRequest, Response> primary;
+    final ReplicasProxy<ReplicaRequest> replicasProxy;
+    final AtomicBoolean finished = new AtomicBoolean();
+    final ActionListener<Response> finalResponseListener;
+
+    volatile Response finalResponse = null;
+
+    private final List<ReplicationResponse.ShardInfo.Failure> shardReplicaFailures = Collections.synchronizedList(new ArrayList<>());
+
+
+    public ReplicatedOperation(Request request, Primary<Request, ReplicaRequest, Response> primary,
+                               ActionListener<Response> listener,
+                               boolean executeOnReplicas, boolean checkWriteConsistency,
+                               ReplicasProxy<ReplicaRequest> replicasProxy,
+                               Supplier<ClusterState> clusterStateSupplier, ESLogger logger, String opType) {
+        this.checkWriteConsistency = checkWriteConsistency;
+        Objects.requireNonNull(request.shardId(), "missing shardId");
+        this.executeOnReplicas = executeOnReplicas;
+        this.replicasProxy = replicasProxy;
+        this.primary = primary;
+        this.finalResponseListener = listener;
+        this.logger = logger;
+        this.request = request;
+        this.clusterStateSupplier = clusterStateSupplier;
+        this.opType = opType;
+    }
+
+
+    // nocommit: deal with exceptions, tricky part is exception during replication logic where operation has already been performed locally
+    public void performOnPrimary() throws Exception {
+
+        final String writeConsistencyFailure = checkWriteConsistency ? checkWriteConsistency() : null;
+        if (writeConsistencyFailure != null) {
+            finishAsFailed(new UnavailableShardsException(
+                    request.shardId(), "{} Timeout: [{}], request: [{}]", writeConsistencyFailure, request.timeout(), request));
+            return;
+        }
+
+        pendingShards.incrementAndGet(); // increase by 1 until we finish all primary coordination
+        Tuple<Response, ReplicaRequest> primaryResponse = primary.performOperation(request);
+        finalResponse = primaryResponse.v1();
+        ReplicaRequest replicaRequest = primaryResponse.v2();
+        replicaRequest.primaryTerm(primary.term());
```

Comment:
This could be done in primary.performOperation(request). No need to expose primary terms here (also remove it from Primary interface).